### PR TITLE
Migrated `pod-security-admission` to contextual logging

### PR DESCRIPTION
**Pull request by Namanl2001**
Signed-off-by: Naman <namanlakhwani@gmail.com>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
It migrates pod-security-admission to contextual and structured logging which is superior to the current unstructured logging.
#### Which issue(s) this PR fixes:
Part of https://github.com/kubernetes/enhancements/issues/3077

#### Does this PR introduce a user-facing change?

```release-note
Migrated `pod-security-admission` to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

/wg structured-logging
/cc @pohly @shivanshu1333 
/cc @kubernetes/wg-structured-logging-reviews 
/priority important-longterm
/area logging

### CODE REVIEW

File: staging/src/k8s.io/pod-security-admission/admission/admission.go
Comment (pohly) [inline staging/src/k8s.io/pod-security-admission/admission/admission.go:230]: `logger` is only used in error paths. Instead of paying the price for this call here on all calls to `ValidateNamespace`, better repeat it whenever needed.

File: staging/src/k8s.io/pod-security-admission/admission/admission.go
Comment (pohly) [inline staging/src/k8s.io/pod-security-admission/admission/admission.go:559]: Look up the logger here.

File: staging/src/k8s.io/pod-security-admission/admission/admission.go
Comment (pohly) [inline staging/src/k8s.io/pod-security-admission/admission/admission.go:331]: Same here.

**Review by pohly — CHANGES_REQUESTED** [review]

File: staging/src/k8s.io/pod-security-admission/admission/admission.go
Comment (Namanl2001) [reply staging/src/k8s.io/pod-security-admission/admission/admission.go]: Makes sense. 

Done.

**Review by pohly — APPROVED** [review]
/lgtm
/label tide/merge-method-squash

If you want to squash manually (leads to a nicer commit), then can also LGTM again.

**Comment by Namanl2001** [comment]
> If you want to squash manually (leads to a nicer commit)

done

@pohly - do we need to make changes in [admission_test.go](https://github.com/kubernetes/kubernetes/blob/74800100c719f2d7859b4cd44ca9977d59a2c778/staging/src/k8s.io/pod-security-admission/admission/admission_test.go) as well? like mentioned [here](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/3077-contextual-logging/README.md#unit-testing). How to recognize when changes in unit tests are needed and when not?

**Comment by pohly** [comment]
That's a good point. Because you didn't have to change function signatures, the unit tests continue to work unchanged and therefore changes there are not really needed. But doing them anyway provides some benefits like per-test log output, so it makes sense to also update them.

**Review by pohly — APPROVED** [review]
/lgtm

**Comment by sftim** [comment]
I think this does need a release note, as end users could observe this change.

**Comment by Namanl2001** [comment]
> I think this does need a release note, as end users could observe this change.

sure, please check if the current note sounds good!

**Review by pohly — APPROVED** [review]
/lgtm
/assign @soltysh 

For approval.

**Comment by stlaz** [comment]
/triage accepted
/assign @tallclair 
for review

**Comment by soltysh** [comment]
/unassign
this goes through sig-auth, also @Namanl2001 can you please rebase your PR?

**Comment by Namanl2001** [comment]
last commit (`calling logger once`) is in favour of https://github.com/kubernetes/kubernetes/pull/114918/files#r1127472443

File: staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go
Comment (tallclair) [inline staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go:115]: This is essentially the main package for the server. You can see where the context is propagated through on line 71. Is there value to migrating this package to contextual logging, if you're not adding a logger to the context?

File: staging/src/k8s.io/pod-security-admission/admission/admission.go
Comment (tallclair) [inline staging/src/k8s.io/pod-security-admission/admission/admission.go:471]: nit: inline this in the if to keep it scoped here
```suggestion
	if klogV := logger.V(5); klogV.Enabled() {
```

File: staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go
Comment (pohly) [reply staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go:115]: It helps with tracking which code still needs to be converted: if we don't convert server.go, then we need to keep it in the exception list. It's simpler to convert it.

File: staging/src/k8s.io/pod-security-admission/admission/admission.go
Comment (pohly) [reply staging/src/k8s.io/pod-security-admission/admission/admission.go:471]: That is indeed the recommended pattern.

File: staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go
Comment (pohly) [reply staging/src/k8s.io/pod-security-admission/cmd/webhook/server/server.go:115]: @tallclair: are you okay with converting this?

LGTM, but I'll defer to @tallclair. We also still need approval.

**Pull request merged**
